extern crate log;

use rand::seq::IndexedRandom;
use rand::Rng;
use std::collections::{HashMap, VecDeque};
use std::io;
use std::io::Write;
use common::file_tools::open_file;
use common::neat_rng::NeatRng;
use common::structs::nucleotides::{Nuc, sequence_array_to_string};
use common::structs::variants::Variant;

fn genotype_to_string(genotype: &Vec<u8>) -> String {
    // Converts a vector of 0s and 1s representing genotype to a standard
    // vcf genotype string.
    let mut geno_string = String::new();
    for ploid in genotype {
        geno_string += &format!("{}/", *ploid.to_string())
    }
    geno_string.strip_suffix("/").unwrap().to_string()
}

pub fn write_vcf(
    variant_locations: &Box<HashMap<String, HashMap<usize, Variant>>>,
    fasta_order: &VecDeque<String>,
    fasta_lengths: &HashMap<String, usize>,
    reference_path: &str,
    overwrite_output: bool,
    output_file_prefix: &str,
) -> io::Result<()> {
    // Takes:
    // variant_locations: A map of contig names keyed to lists of variants in that contig
    // consisting of a tuple of (position, alt base, ref base).
    // fasta_order: A vector of contig names in the order of the reference fasta.
    // ploidy: The number of copies of each chromosome present in the organism
    // reference_path: The location of the reference file this vcf is showing variants from.
    // output_file_prefix: The path to the directory and the prefix to use for filenames
    // rng: A random number generator for this run
    // Result:
    // Throws and error if there's a problem, or else returns nothing.
    //
    // set the filename of the output vcf
    let mut filename = format!("{}.vcf", output_file_prefix);
    let mut outfile = open_file(&mut filename, overwrite_output)
        .expect(&format!("Problem opening {} for output.", filename));
    // add the vcf header
    writeln!(&mut outfile, "##fileformat=VCFv4.1")?;
    writeln!(&mut outfile, "##reference={}", reference_path)?;
    for contig in &fasta_order {
        writeln!(&mut outfile, "##contig=<ID={},length={}>", contig, fasta_lengths.get(&contig).unwrap().to_string())?;
    }
    writeln!(&mut outfile, "##Generated by rusty-neat")?;
    writeln!(
        &mut outfile,
        "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Total Depth\">"
    )?;
    writeln!(
        &mut outfile,
        "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency\">"
    )?;
    writeln!(&mut outfile, "##INFO=<ID=VMX,Number=1,Type=String, Description=\"SNP is Missense in these Read Frames\">")?;
    writeln!(&mut outfile, "##INFO=<ID=VNX,Number=1,Type=String, Description=\"SNP is Nonsense in these Read Frames\">")?;
    writeln!(
        &mut outfile,
        "##INFO=<ID=VFX,Number=1,Type=String,Description=\"Indel Causes Frameshift\">"
    )?;
    writeln!(&mut outfile, "##ALT=<ID=DEL,Description=\"Deletion\">")?;
    writeln!(&mut outfile, "##ALT=<ID=DUP,Description=\"Duplication\">")?;
    writeln!(
        &mut outfile,
        "##ALT=<ID=INS,Description=\"Insertion of novel sequence\">"
    )?;
    writeln!(&mut outfile, "##ALT=<ID=INV,Description=\"Inversion\">")?;
    writeln!(
        &mut outfile,
        "##ALT=<ID=CNV,Description=\"Copy number variable region\">"
    )?;
    writeln!(
        &mut outfile,
        "##ALT=<ID=TRANS,Description=\"Translocation\">"
    )?;
    writeln!(
        &mut outfile,
        "##ALT=<ID=INV-TRANS,Description=\"Inverted translocation\">"
    )?;
    writeln!(
        &mut outfile,
        "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">"
    )?;
    // Add a neat sample column
    writeln!(
        &mut outfile,
        "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tNEAT_simulated_sample"
    )?;
    // insert mutations
    for contig in fasta_order {
        for (position, mutation) in &variant_locations[contig] {
            // Format the output line. Any fields without data will be a simple period. Quality
            // is set to 37 for all these variants.
            let line = format!(
                "{}\t{}\t.\t{}\t{}\t37\tPASS\t.\tGT\t{}",
                contig,
                position + 1,
                sequence_array_to_string(mutation.reference),
                sequence_array_to_string(mutation.alternate),
                genotype_to_string(&mutation.genotype),
            );
            writeln!(&mut outfile, "{}", line)?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use std::fs;
    use std::path::Path;
    use common::neat_rng::NeatRng;
    use common::structs::nucleotides::Nuc::*;
    use common::structs::variants::{Variant, VariantType};

    #[test]
    fn test_genotype_to_string() {
        let genotype = vec![0, 1, 0];
        assert_eq!(String::from("0/1/0"), genotype_to_string(&genotype));
    }

    #[test]
    fn test_write_vcf() {
        let variant_locations = Box::from(HashMap::from([
            (
                "chr1".to_string(), 
                HashMap::from([
                    (3, Variant::new(
                        VariantType::SNP,
                        &vec![A],
                        &vec![C],
                        vec![1, 0],
                        false)
                    ),
                    (7, Variant::new(
                        VariantType::SNP,
                        &vec![T],
                        &vec![G],
                        vec![1, 1],
                        true)
                    )
                ])
            )
        ]));
        let fasta_order = VecDeque::from(["chr1".to_string()]);
        let fasta_length = HashMap::from([("chr 1".to_string(), 20)]);
        let ploidy = 2;
        let reference_path = "/fake/path/to/H1N1.fa";
        let overwrite_output = false;
        let output_file_prefix = "test";
        let mut rng = NeatRng::seed_from_u64(0);
        write_vcf(
            &variant_locations,
            &fasta_order,
            &fasta_length,
            reference_path,
            overwrite_output,
            output_file_prefix,
        )
        .unwrap();
        assert!(Path::new("test.vcf").exists());
        fs::remove_file("test.vcf").unwrap();
    }
}
